cleangoldenami/main.go

```diff
@@ -58,23 +60,60 @@ func main() {
 				log.Fatalf("error fetching images for region %v: %v", regionName, err)
 			}
 
-			if imageCount := len(images); imageCount < quota {
+			if imageCount := len(images); (quota - imageCount) >= desiredImageCapacity {
 				if *verbose {
 					fmt.Printf("Region %v is under quota. (Images: %v, Quota: %v)\n", regionName, imageCount, quota)
 				}
 			} else {
-				imageToDelete, err := getOldestImage(images)
-				if err != nil {
-					log.Fatalf("error determining which image to delete in region %v: %v", regionName, err)
-				}
-				if *dryRun {
-					fmt.Printf("Region %v is at quota (%v) - would delete %v\n", regionName, quota, *imageToDelete.ImageId)
-				} else {
-					err := deregisterImage(ec2Client, imageToDelete)
+				numOfImagesToDelete := desiredImageCapacity - (quota - imageCount)
+
+				arm64Images := filterImages(images, "arm64", "rhel-arm64")
+				legacyx86Images := filterImages(images, "x86_64", "legacy-x86_64")
+				x86Images := filterImages(images, "x86_64", "rhel-x86_64")
+
+				var imagesToDelete []ec2Types.Image
+
+				for !imageDeletionCheck(imagesToDelete, numOfImagesToDelete, arm64Images, legacyx86Images, x86Images) {
+					// Returns the most populated image type and the slice containing images of that type
+					mostPopulatedImagesByType, mostPopulatedImageType := getMostPopulatedImageType(arm64Images, legacyx86Images, x86Images)
+
+					// Returns the oldest image from the most populated type's slice along with its index value
+					ImageToDeleteIndex, ImageToDelete, err := getOldestImage(mostPopulatedImagesByType)
 					if err != nil {
-						log.Fatalf("error deregistering image %v in region %v: %v", *imageToDelete.ImageId, regionName, err)
+						// Log a fatal error if there was an issue retrieving the oldest image
+						log.Fatalf("error determining which image to delete in region %v: %v", regionName, err)
+					}
+
+					// If ImageToDeleteIndex is valid (non-negative), proceed with deletion
+					if ImageToDeleteIndex >= 0 {
```

Comment:
If the index is negative here, that feels like an error we should be logging and doing something with. Maybe we could log an error about the list being empty or something similar.